File: frontend/src/types/privacy.ts
// Types for the CCPA data export / deletion flow — docs/API_CONTRACTS.md
// "Privacy (CCPA data export / deletion)". Backend (PR #70) has been Done
// with zero frontend consumer until this page — matches backend/app/
// schemas/privacy.py exactly.

export interface OwnerAccountExport {
  id: number;
  cognito_sub: string;
  email: string;
  full_name: string | null;
  phone: string | null;
  stripe_customer_id: string | null;
  created_at: string;
  personal_data_deleted_at: string | null;
}

export interface LocationManagerExport {
  location_id: number;
  is_active: boolean;
  assigned_at: string;
  revoked_at: string | null;
}

export interface FollowExport {
  brand_id: number;
  followed_at: string;
}

export interface ClaimRequestExport {
  claim_id: number;
  brand_id: number;
  status: string;
  proof_method: string;
  submitted_at: string;
  reviewed_at: string | null;
}

/**
 * A "report a problem" submission attributed to the caller — matched by
 * `listing_report.reporter_user_id` (their Cognito `sub`), not by
 * `reporter_email` (free text any submitter, including an anonymous one,
 * can type — see backend `ListingReportExportOut` docstring).
 */
export interface ListingReportExport {
  report_id: number;
  brand_id: number;
  location_id: number | null;
  category: string;
  details: string;
  reporter_email: string | null;
  status: string;
  submitted_at: string;
  reviewed_at: string | null;
}

/**
 * A search or restaurant-tile click recorded while signed in as a
 * registered user (12-month retention; hard-deleted by a deletion request).
 * `payload` is exactly what was stored — kept loose since it differs by
 * `event_type` and is only ever shown in the raw JSON export.
 */
export interface ActivityEventExport {
  event_type: string;
  created_at: string;
  payload: Record<string, unknown>;
}

/** Actions the caller themselves performed — retained, never touched by a deletion request. */
export interface AuditLogExport {
  table_name: string;
  record_id: number;
  action: string;
  actor_role: string;
  created_at: string;
}

/**
 * The display name + last-seen record a registered user / manager has (the
 * counterpart of `owner_account` for those roles). `null` on `DataExport`
 * when no row exists. Hard-deleted by an approved deletion request.
 */
export interface UserProfileExport {
  full_name: string | null;
  last_seen_at: string | null;
  updated_at: string;
}

/** Full response for `GET /auth/me/data-export` — synchronous, JSON. */
export interface DataExport {
  cognito_sub: string;
  role: string;
  email: string | null;
  generated_at: string;
  owner_account: OwnerAccountExport | null;
  user_profile: UserProfileExport | null;
  location_manager_assignments: LocationManagerExport[];
  follows: FollowExport[];
  claim_requests: ClaimRequestExport[];
  listing_reports: ListingReportExport[];
  activity_events: ActivityEventExport[];
  audit_log_entries: AuditLogExport[];
  notice: string;
}

/** Body for `POST /auth/me/data-deletion`. */
export interface CreateDataDeletionInput {
  reason?: string | null;
}

/**
 * `data_scope` on `DataDeletionRequest` is a row-count snapshot object
 * (`{ owner_account: 0, follows: 3, ... }`) — kept loose here (matching the
 * backend's `dict[str, Any] | None`) since it's admin-review context, not
 * something this page renders field-by-field.
 */
export type DataDeletionScope = Record<string, number> | null;

/** One request row, shared by POST /auth/me/data-deletion's response and GET's list. */
export interface DataDeletionRequest {
  request_id: number;
  status: string;
  requester_role: string;
  reason: string | null;
  data_scope: DataDeletionScope;
  submitted_at: string;
  reviewed_at: string | null;
  reviewer_notes: string | null;
  completed_at: string | null;
}

File: backend/app/schemas/privacy.py
"""CCPA data export / deletion — see docs/API_CONTRACTS.md "Privacy (CCPA
data export / deletion)"."""
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel


class OwnerAccountExportOut(BaseModel):
    id: int
    cognito_sub: str
    email: str
    full_name: str | None
    phone: str | None
    stripe_customer_id: str | None
    created_at: datetime
    personal_data_deleted_at: datetime | None


class LocationManagerExportOut(BaseModel):
    location_id: int
    is_active: bool
    assigned_at: datetime
    revoked_at: datetime | None


class FollowExportOut(BaseModel):
    brand_id: int
    followed_at: datetime


class ClaimRequestExportOut(BaseModel):
    claim_id: int
    brand_id: int
    status: str
    proof_method: str
    submitted_at: datetime
    reviewed_at: datetime | None


class ListingReportExportOut(BaseModel):
    """A "report a problem" submission attributed to the caller — matched
    by `listing_report.reporter_user_id` (their Cognito `sub`), NOT by
    `reporter_email`: that field is free text any submitter (including an
    anonymous one) can type, so it is not a reliable identity match — see
    `app/services/privacy_service.py` module docstring.
    """

    report_id: int
    brand_id: int
    location_id: int | None
    category: str
    details: str
    reporter_email: str | None
    status: str
    submitted_at: datetime
    reviewed_at: datetime | None


class AuditLogExportOut(BaseModel):
    """Actions the caller themselves performed, per DECISIONS.md "CCPA
    data export/deletion" — included for transparency, but retained (not
    deleted/anonymized) by a data-deletion request; see `notice` on
    `DataExportOut`.
    """

    table_name: str
    record_id: int
    action: str
    actor_role: str
    created_at: datetime


class DataExportOut(BaseModel):
    cognito_sub: str
    role: str
    email: str | None
    generated_at: datetime
    owner_account: OwnerAccountExportOut | None
    location_manager_assignments: list[LocationManagerExportOut]
    follows: list[FollowExportOut]
    claim_requests: list[ClaimRequestExportOut]
    listing_reports: list[ListingReportExportOut]
    audit_log_entries: list[AuditLogExportOut]
    notice: str


class DataDeletionRequestCreate(BaseModel):
    reason: str | None = None


class DataDeletionRequestOut(BaseModel):
    request_id: int
    status: str
    requester_role: str
    reason: str | None
    data_scope: dict[str, Any] | None
    submitted_at: datetime
    reviewed_at: datetime | None
    reviewer_notes: str | None
    completed_at: datetime | None


class DataDeletionListResponse(BaseModel):
    results: list[DataDeletionRequestOut]
    page: int
    page_size: int
    total: int


class DataDeletionApproveRequest(BaseModel):
    reviewer_notes: str | None = None


class DataDeletionRejectRequest(BaseModel):
    reviewer_notes: str
